core/object/class_db.cpp

```diff
@@ -377,7 +378,14 @@ bool ClassDB::can_instantiate(const StringName &p_class) {
 	OBJTYPE_RLOCK;
 
 	ClassInfo *ti = classes.getptr(p_class);
-	ERR_FAIL_COND_V_MSG(!ti, false, "Cannot get class '" + String(p_class) + "'.");
+	if (!ti) {
+		if (!ScriptServer::is_global_class(p_class)) {
+			ERR_FAIL_V_MSG(false, "Cannot get class '" + String(p_class) + "'.");
+		}
+		String path = ScriptServer::get_global_class_path(p_class);
+		Ref<Script> scr = ResourceLoader::load(path);
+		return scr.is_valid() && scr->is_valid() && !scr->is_abstract();
+	}
```

Comment:
I guess it makes sense why this is needed, and there is code that checks `ScriptServer::is_global_class()` already in `ClassDB::is_virtual()`, so there's a precedent for handling this sort of thing in `ClassDB`.

But it's not ideal that calling a method on `ClassDB` leads to running `ResourceLoader::load()`. For the most part, `ClassDB` is just updating and querying collections of data, where as this could lead to parsing and compiling a script!

I don't know if this has already been discussed, but is there a way for `ScriptServer` to keep track of which global classes are abstract so that we don't need to invoke the resource loader here?

Reply:
> is there a way for ScriptServer to keep track of which global classes are abstract so that we don't need to invoke the resource loader here?

I discussed with @raulsntos earlier about moving some of this logic to `ScriptServer` for consistency, but I'm not sure it would be possible to check for abstractness without loading the script itself. Especially in the case of C#, [where it needs a loaded script](https://github.com/godotengine/godot/blob/0bca4242392992840b0b891acde6422c49712e3e/modules/mono/csharp_script.cpp#L2295) in order to check for abstractness and other details in the first place. I believe [the resource loader should cache the script anyway](https://github.com/godotengine/godot/blob/0bca4242392992840b0b891acde6422c49712e3e/core/io/resource.cpp#L82), so [it only has to do all the expensive loading once](https://github.com/godotengine/godot/blob/0bca4242392992840b0b891acde6422c49712e3e/core/io/resource_loader.cpp#L465) if it wasn't already loaded in the first place.

I do think it would be nice to have a sort of lightweight script metadata system so that this kind of loading isn't necessary, but that would be out of scope for this PR.

Reply:
> I discussed with @raulsntos earlier about moving some of this logic to `ScriptServer` for consistency, but I'm not sure it would be possible to check for abstractness without loading the script itself.

I think it's fine to load the script _at some point_, to check if it's abstract. What I don't personally like here is that calling a method on `ClassDB` may trigger the loading of the script right then.

Some other mechanism is already telling `ScriptServer` about each global class. Could that same mechanism also tell the `ScriptServer` whether they are abstract or not? Or, maybe the `ScriptLanguage` could periodically check all its global classes and update `ScriptServer` to tell it if they are each abstract or not? We could even trigger that to happen when the "Create node" dialog opens.

Reply:
> Could that same mechanism also tell the ScriptServer whether they are abstract or not?

At the moment, I don't think that this would be possible, or at the very least reasonable for this PR. [This is what is stored about a global class](https://github.com/godotengine/godot/blob/0bca4242392992840b0b891acde6422c49712e3e/core/object/script_language.h#L59-L63) in `ScriptServer` -- the bare minimum. In order to get anything else about a class (validity, tool-ness, abstractness, etc.) a call to `ResourceLoader::load` is required. I've tried to find the code which adds classes to the global class list and the best I could find is in [the editor file system](https://github.com/godotengine/godot/blob/0bca4242392992840b0b891acde6422c49712e3e/editor/editor_file_system.cpp#L1548C30-L1548C30), though that could very well be my editor being lazy and me not fully comprehending this system. In any case, that's not the focus of this PR and even if it was, I do not see a simple solution to passing arbitrary script information like this to `ScriptServer`. I would love to explore this idea further in future PRs, however.

> Or, maybe the `ScriptLanguage` could periodically check all its global classes and update `ScriptServer` to tell it if they are each abstract or not?

This would also require a `ResourceLoader::load` since a `ScriptLanguage` does not keep track of global classes on its own (see [C#](https://github.com/godotengine/godot/blob/0bca4242392992840b0b891acde6422c49712e3e/modules/mono/csharp_script.cpp#L550-L586) and [GDScript](https://github.com/godotengine/godot/blob/0bca4242392992840b0b891acde6422c49712e3e/modules/gdscript/gdscript.cpp#L2531-L2633) implementations of `ScriptLanguage::get_global_class_name`). This would instead lead to every single script being loaded within the project periodically instead of only when needed, which I do not think is a good optimization.

> We could even trigger that to happen when the "Create node" dialog opens.

I believe the `CreateDialog` [already calls `ResourceLoader::load`](https://github.com/godotengine/godot/blob/0bca4242392992840b0b891acde6422c49712e3e/editor/create_dialog.cpp#L221-L238) on script classes, and because it builds a tree of every type the engine is aware of, I think it's safe to say that it already does this. The problem with this idea though, is that this needs to happen every time we want to check for abstractness. Which means every script must _also_ be loaded in, for example, `EditorResourcePicker`, [where we only want to check abstractness for a single type](https://github.com/godotengine/godot/blob/0bca4242392992840b0b891acde6422c49712e3e/editor/editor_resource_picker.cpp#L496-L527).

I personally do not see an issue with calling `ResourceLoader::load` here in `ClassDB` since most likely the result is already going to be cached. If it isn't, that just means a few microseconds of load time saved in the near future. Besides, this is load call only ever called on global classes, which means that built in types and non-global scripts _never_ have to deal with a random resource load.

Reply:
I think you're misunderstanding what I'm suggesting. Using `ResourceLoader::load()` is fine to determine if the class is abstract. I'm just saying that I think it'd be better if wasn't triggered by calling a `ClassDB` function. Running a function before you call `ClassDB::can_instantiate()` that updates which global classes are abstract (by using the resource loader) would be fine.

In any case, if the core team is fine with calling a `ClassDB` function triggering the resource loader than it's fine! I just personally think it muddies `ClassDB` as a neutral data store that all parts of Godot can interact with while remaining decoupled.

Reply:
Ohh okay, I think I better understand now.
> I think it'd be better if wasn't triggered by calling a `ClassDB` function . . . it muddies `ClassDB` as a neutral data store that all parts of Godot can interact with while remaining decoupled.

I 100% agree here, but I'm not sure how else it could be implemented at this time without creating a massive headache. This is something I will try to look into in the future, though.

For now, however, it seems like the core team is fine with the current implementation and it appears to work correctly, so I'm going to go ahead and mark this as resolved unless there are any other thoughts.

Reply:
For #67777 I ended up moving the logic into CreateDialog instead of ClassDB, but I'm not sure what is the best place for it.

Reply:
I would probably use something like `ScriptServer` since this would be used throughout various unrelated parts of the engine. The only issue is that `ClassDB::can_instantiate` and `ClassDB::is_virtual` both have the same fallback return value, `false`, which means that an invalid script could lead to both being returning `false`. Trying to centralize this into a single function would mean that the fallback return value now differs if it were defined like so:

```cpp
bool ScriptServer::is_abstract(const StringName &p_class) {
    if (!ScriptServer::is_global_class(p_class)) {
        ERR_FAIL_V_MSG(false, "Cannot get class '" + String(p_class) + "'.");
    }
    String path = ScriptServer::get_global_class_path(p_class);
    Ref<Script> scr = ResourceLoader::load(path);
    return scr.is_valid() && scr->is_valid() && scr->is_abstract();
}

// ClassDB::can_instantiate
if (!ti) {
    // Returns true on invalid script
    return !ScriptServer::is_abstract(p_class);
}

// ClassDB::is_virtual
if (!ti) {
    // Returns false on invalid script
    return ScriptServer::is_abstract(p_class);
}
```

Which I suppose then means that there should be _two_ `ScriptServer` functions (`can_instantiate` and `is_virtual`/`is_abstract`) or some sort of boolean return parameter which is only `true` when an error is called, but I don't really like either solution.

This could probably allow for that `ResourceLoader::load` call to be removed from those `ClassDB` functions, but since these two sets of functions would most likely always be called together, I more see this as unnecessary refactoring.

Reply:
What about something like this?

```cpp
bool ScriptServer::is_abstract(const String &p_type) const {
	if (ClassDB::class_exists(p_type)) {
		return !ClassDB::can_instantiate(p_type) || ClassDB::is_virtual(p_type);
	}
	if (ScriptServer::is_global_class(p_type)) {
		String path = ScriptServer::get_global_class_path(p_type);
		Ref<Script> scr = ResourceLoader::load(path, "Script");
		return scr->is_abstract();
	}
	return false;
}
```

Reply:
I don't know if `ScriptServer` should be used like this since it only manages script types, not built-in types. However, I do think this is a good idea. Only thing I would do is move the `ClassDB` checks to be run when `ScriptServer::is_global_class` returns `false` as to avoid unnecessary `ClassDB` searching. If the global class code is moved out of those `ClassDB` functions, an error would be thrown when `p_type` is not a global or built-in type.

Again, though, this would then require all `ClassDB::can_instantiate` usages to be modified to use this new function in order to retain functionality (except in cases where only built-in types are meant to be handled, but I'm not sure when or even if that would be necessary), which feels a little out of scope for this PR.

Reply:
Then it seems like having this be in ClassDB is the right call then.